### PR TITLE
plan: add task profile.

### PR DESCRIPTION
**Pull request by hanfei1991**
We are planning to split the physical plan to computation tasks. This pr only define the structure for kvTask and singleton Task.

### CODE REVIEW

**Comment by hanfei1991** [comment]
@shenli @coocood @zimulala @tiancaiamao @XuHuaiyu @ilovesoup @lamxTyler @winoros 
PTAL

**Comment by coocood** [comment]
This need a complete design document.

**Comment by shenli** [comment]
@coocood I am agree with you.

**Comment by coocood** [comment]
s/kvTask/copTask
s/singletonTask/rootTask

File: plan/task_profile.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import "github.com/pingcap/tidb/expression"
+
+// taskProfile is a new version of `PhysicalPlanInfo`. It stores cost information for a task.
+// A task may be KVTask, SingletonTask, MPPTask or a ParallelTask.
```
Comment (shenli) [inline plan/task_profile.go]: Update the comment.

File: plan/task_profile.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+type taskProfile interface {
```
Comment (shenli) [inline plan/task_profile.go]: Need a toPB interface?

Comment (hanfei1991) [inline plan/task_profile.go]: Nope, toPB only required by kvTask.

File: plan/task_profile.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	cost() float64
+	copy() taskProfile
+}
+
+// TODO: In future, we should split copTask to indexTask and tableTask.
+// copTaskProfile is a profile for a task running a distributed kv store.
+type copTaskProfile struct {
+	indexPlan     PhysicalPlan
+	tablePlan     PhysicalPlan
+	cst           float64
```
Comment (shenli) [inline plan/task_profile.go]: Why not use cost? cst is rarely used.

File: plan/task_profile.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		p.SetChildren(nt.tablePlan)
+		nt.tablePlan = p
+	}
+	return nt
+}
+
+func (t *copTaskProfile) finishTask() {
+	t.cst += float64(t.cnt) * netWorkFactor
+	if t.tablePlan != nil && t.indexPlan != nil {
+		t.cst += float64(t.cnt) * netWorkFactor
```
Comment (shenli) [inline plan/task_profile.go]: t.cst *= 2？

File: plan/task_profile.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// TODO: In future, we should split copTask to indexTask and tableTask.
+// copTaskProfile is a profile for a task running a distributed kv store.
+type copTaskProfile struct {
+	indexPlan     PhysicalPlan
+	tablePlan     PhysicalPlan
+	cst           float64
+	cnt           uint64
+	addPlan2Index bool
```
Comment (shenli) [inline plan/task_profile.go]: Add comment for this.

File: plan/task_profile.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		}
+		t.cnt = uint64(float64(t.cnt) * selectionFactor)
+	case *rootTaskProfile:
+		t.cst += float64(t.cnt) * cpuFactor
+		t.cnt = uint64(float64(t.cnt) * selectionFactor)
+	}
+	return profile
+}
+
+func (sel *Selection) splitSelectionByIndexColumns(schema *expression.Schema) (indexSel *Selection, tableSel *Selection) {
```
Comment (shenli) [inline plan/task_profile.go]: Add comment for this function.

File: plan/task_profile.go (hunk 3 above)
Comment (hanfei1991) [inline plan/task_profile.go]: cost is used in function name

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM
@coocood @zimulala @tiancaiamao PTAL

File: plan/task_profile.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+// limitations under the License.
+
+package plan
+
+import "github.com/pingcap/tidb/expression"
+
+// taskProfile is a new version of `PhysicalPlanInfo`. It stores cost information for a task.
+// A task may be CopTask, RootTask, MPPTask or a ParallelTask.
+type taskProfile interface {
+	attachPlan(p PhysicalPlan) taskProfile
```
Comment (coocood) [inline plan/task_profile.go]: Use a function instead of interface method is more intuitive.

**Pull request merged**